Kuadrant/wasm-shim
# New pluginconfiguration structure and implementation

### What

Implements partially https://github.com/Kuadrant/kuadrant-controller/issues/129

New PluginConfiguration structure. 

Each rate limit policy has a logical name and a set of hostnames to activate it based on the incoming request’s host header.

The WASM-SHIM builds a tree based data structure holding the rate limit policies. The longest (sub)domain match is used to select the policy to be applied. Only one policy is being applied per invocation.

The plugin object contains a list of configurations to build a list of Envoy's RLS descriptors. These policy configurations are defined at

```
rate_limit_policies[*].gatewat_actions[*].configurations
```

For example:

```yaml
configurations:
- actions:
   - generic_key:
        descriptor_key: admin
        descriptor_value: "1"
```

Each configuration produces, at most, one descriptor. Depending on the incoming request, one configuration may or may not produce a rate limit descriptor.

Each policy configuration has associated, optionally, a set of rules to match. Rules allow matching `hosts` and/or `methods` and/or `paths`. Matching occurs when at least one rule applies against the incoming request. If rules are not set or empty, it is equivalent to matching all the requests.

For example:

```yaml
gateway_actions:
      - rules:
          - paths: ["/admin/toy"]
            methods: ["GET"]
            hosts: ["pets.toystore.com"]
        configurations:
          - actions:
            - generic_key:
                descriptor_key: admin
                descriptor_value: "1"
```

Each configuration object defines a list of actions. Each action may (or may not) produce a descriptor entry (descriptor list item).
If an action cannot append a descriptor entry, no descriptor is generated for the configuration.

The external rate limit service will be called when there is at least one not empty descriptor generated by the list of configurations.

Example of plugin configuration object

```yaml
failure_mode_deny: true
rate_limit_policies:
  - name: toystore
    rate_limit_domain: toystore-app
    upstream_cluster: rate-limit-cluster
    hostnames: ["*.toystore.com"]
    gateway_actions:
      - rules:
          - paths: ["/admin/toy"]
            methods: ["GET"]
            hosts: ["pets.toystore.com"]
        configurations:
          - actions:
            - generic_key:
                descriptor_key: admin
                descriptor_value: "1"
```

### Verification steps

* Start local testing env based on docker-compose
```
$ make development
```

Test rate limit policy `rlp-a`. The WASM filter should not call limitador (reason: actions should not genereate descriptors)

```
curl -H "Host: test.a.com" http://127.0.0.1:18000/get
```

Test rate limit policy `rlp-b`. The WASM filter should not call limitador (reason: rules do not match)

```
curl -H "Host: test.b.com" http://127.0.0.1:18000/get
```

Test rate limit policy `rlp-c`. The WASM filter should call limitador with 4 descriptors

```
curl -H "Host: test.c.com" -H "x-forwarded-for: 127.0.0.1" -H "My-Custom-Header-01: my-custom-header-value-01" -H "My-Custom-Header-02: my-custom-header-value-02" http://127.0.0.1:18000/get
```

The Limitador's logs should show RLS request with 4 descriptors

```
limitador_1  | [2022-06-27T23:50:33Z DEBUG limitador_server::envoy_rls::server] Request received: Request { metadata: MetadataMap { headers: {"te": "trailers", "grpc-timeout": "5000m", "content-type": "application/grpc", "x-envoy-internal": "true", "x-forwarded-for": "172.20.0.4", "x-envoy-expected-rq-timeout-ms": "5000"} }, message: RateLimitRequest { domain: "rls-domain-c", descriptors: [RateLimitDescriptor { entries: [Entry { key: "some_generic_key_for_c", value: "some_generic_value_for_c" }], limit: None }, RateLimitDescriptor { entries: [Entry { key: "remote_address", value: "127.0.0.1" }], limit: None }, RateLimitDescriptor { entries: [Entry { key: "header_match", value: "header_value_match_value" }], limit: None }, RateLimitDescriptor { entries: [Entry { key: "my-custom-header-key", value: "my-custom-header-value-02" }], limit: None }], hits_addend: 1 }, extensions: Extensions }
```

PR's added to the `new-design` branch:

* Policy index tree #14 
* New plugin config struct #19 
* HTTP context implementation with the new filter config struct #20 
*  README and local testing resources #21
